FTN/ring_buffer.c

```diff
@@ -0,0 +1,167 @@
+    }
+    if (RB_is_full(ring_buff))
+    {
+        printf("Queue is Full, couldnt write\n");
+        return false;
+    }
+
+    memcpy(&ring_buff->msgs[ring_buff->tail], new_msg, sizeof(msg));
+    if (CLIENTS[GET_PRIVATE_ID].id == 1){
+        printf("count: %ld <------------------: head: %ld, tailg got insert: %ld\n", ring_buff->count + 1, ring_buff->head, ring_buff->tail);
+        DumpHex((ring_buff->msgs[ring_buff->tail]).msg, (ring_buff->msgs[ring_buff->tail]).len); // <<====
+        printf(" <----------------------------------\n");
+    }
+    ring_buff->tail++;
+
+    if (ring_buff->tail == RING_BUFFER_SIZE) // end of circular buffer?
+    {
+        ring_buff->tail = 0; // turn around
+    }
+
+    ring_buff->count++;
+
+    return true;
+}
+
+bool RB_extract(ring_buffer *ring_buff, msg *out_msg)
+{
+    if (RB_empty(ring_buff))
+    {
+        printf("RB_extract- Queue is Empty\n");
+        return false;
+    }
+    memcpy(out_msg, &ring_buff->msgs[ring_buff->head], sizeof(msg));
+
+    printf("%ld ------------------>: head of extract: %ld, tail: %ld\n", e++, ring_buff->head, ring_buff->tail);
+    DumpHex((ring_buff->msgs[ring_buff->head]).msg, (ring_buff->msgs[ring_buff->head]).len); // <<====
+    printf(" ---------------------------------->\n");
+
+    ring_buff->head++;
+
+    printf("thread couses: RB_extract: tail- %ld, head- %ld, len of stored data: %ld, seq_num: %ld, total in this cli: %ld\n", ring_buff->tail, ring_buff->head, ring_buff->msgs[ring_buff->tail].len, ring_buff->msgs[ring_buff->tail].seq_num, ring_buff->count + 1);
+    if (ring_buff->head == RING_BUFFER_SIZE) // end of circular buffer?
+    {
+        printf("***********\n");
+        ring_buff->head = 0; // turn around
+    }
+
+    ring_buff->count--;
+    return true;
+}
+*/
+bool RB_peek(ring_buffer *ring_buff, uint32_t *out_index)
+{
+    if (RB_empty(ring_buff))
+    {
+        // printf("RB_peek- Queue is Empty\n");
+        return false;
+    }
+
+    *out_index = ring_buff->head % RING_BUFFER_SIZE;
+
+    return true;
+}
+
+bool RB_empty(ring_buffer *ring_buff)
+{
+    return ring_buff->count == 0;
+}
+
+bool RB_extract_first(ring_buffer *ring_buff, uint64_t num_of_rings, msg *out_oldest_msg, uint64_t *opt_out_source_address_id)
+{
+    uint64_t i = 0;
+    uint32_t curr_i_head = 0;
+    uint64_t cli_with_oldest_msg = 0;
+    uint64_t oldest_msg_seq_num = 0xFFFFFFFFFFFFFFFF;
+
+    for (i = SERVER_ADDRESS_ID; i < num_of_rings; i++)
+    {
+        if (RB_peek(&(ring_buff[i]), &curr_i_head))
+        {
+            if (ring_buff[i].msgs[curr_i_head].seq_num < oldest_msg_seq_num)
+            {
+                oldest_msg_seq_num = ring_buff[i].msgs[curr_i_head].seq_num;
+                cli_with_oldest_msg = i;
+            }
+        }
+    }
+    if (0 == cli_with_oldest_msg)
+    {
+        return false; // oldest message not found - no messages in the rings
+    }
+    RB_extract(&ring_buff[cli_with_oldest_msg], out_oldest_msg);
```

Comment:
BUG: time of check time of use on `cli_with_oldest_msg`. maybe multiple callers to this function got the same `cli_with_oldest_msg `